src/network/requestQueue/index.js

```diff
@@ -0,0 +1,117 @@
+const SocketRequest = require('./socketRequest')
+
+module.exports = class RequestQueue {
+  /**
+   * @param {number} maxInFlightRequests
+   * @param {number} requestTimeout
+   * @param {string} clientId
+   * @param {string} broker
+   * @param {Object} logger
+   */
+  constructor({ maxInFlightRequests, requestTimeout, clientId, broker, logger }) {
+    this.maxInFlightRequests = maxInFlightRequests
+    this.requestTimeout = requestTimeout
+    this.clientId = clientId
+    this.broker = broker
+    this.logger = logger
+
+    this.inflight = new Map()
+    this.pending = []
+  }
+
+  createRequest({ entry, expectResponse, send: sendRequest }) {
+    const request = new SocketRequest({
+      entry,
+      expectResponse,
+      broker: this.broker,
+      requestTimeout: this.requestTimeout,
+      send: () => {
+        this.inflight.set(entry.correlationId, request)
+        sendRequest()
+      },
+    })
+
+    return request
+  }
+
+  push(request) {
```

Comment:
Currently we advance through the queue essentially by chaining callbacks - when one requests succeeds/fails we add the next pending message to the queue. If for whatever reason an error or bug prevented the expected end condition from being met we could encounter some sort of a deadlock. I do think that in most cases the timeout will account for this, but it does make me wonder if it'd make sense to have some sort of "cleanup" process running on a setInterval to ensure the queue is advancing. Just a thought, I don't think it's necessarily something we have to implement now.

Reply:
This case is covered, besides the timeout, all requests are rejected on socket errors/timeout (https://github.com/tulios/kafkajs/blob/master/src/network/connection.js#L116), the request queue will reject all inflight and pending requests.